src/core/shape_inference/include/col2im_shape_inference.hpp

```diff
@@ -0,0 +1,125 @@
+// Copyright (C) 2018-2024 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+
+#include <cmath>
+
+#include "openvino/op/col2im.hpp"
+#include "utils.hpp"
+
+namespace ov {
+namespace op {
+namespace v15 {
+template <class TShape, class TRShape = result_shape_t<TShape>>
+std::vector<TRShape> shape_infer(const Col2Im* op,
+                                 const std::vector<TShape>& input_shapes,
+                                 const ITensorAccessor& tensor_accessor = make_tensor_accessor()) {
+    NODE_SHAPE_INFER_CHECK(op,
+                           input_shapes,
+                           input_shapes.size() == 3,
+                           "Number of inputs has to be equal to 3. Got: ",
+                           input_shapes.size());
+
+    const auto& data_shape = input_shapes[0];
+    const auto& output_size_shape = input_shapes[1];
+    const auto& kernel_shape = input_shapes[2];
+    const bool is_batched = data_shape.rank() == 3;
+    const auto output_size_val = ov::op::get_input_const_data_as<TRShape, int64_t>(op, 1, tensor_accessor);
+    const auto kernel_val = ov::op::get_input_const_data_as<TRShape, int64_t>(op, 2, tensor_accessor);
+
+    NODE_SHAPE_INFER_CHECK(op,
+                           input_shapes,
+                           data_shape.rank().compatible(2) || data_shape.rank().compatible(3),
+                           "input data must be an unbatched 2D or a batched 3D input. Got: ",
+                           data_shape);
+
+    if (output_size_shape.rank().is_static()) {
+        NODE_SHAPE_INFER_CHECK(op,
+                               input_shapes,
+                               output_size_shape.rank().compatible(1) && output_size_shape[0].compatible(2),
+                               "output_size must be a 1D input of shape [2]. Got: ",
+                               output_size_shape);
+    }
```

Comment:
This check means that `output_size_shape` can be {2} or {-1} or {{0,2}}, is that OK?

OR only dynamic rank or 2 element 1D shape is allowed?
```suggestion
    const auto is_two_elem_1d = [](const ov::PartialShape& shape) -> bool {
        static const auto exp_shape = ov::PartialShape{2};
        return shape.rank().is_dynamic() || shape == exp_shape;
    };
    NODE_SHAPE_INFER_CHECK(op,
                               input_shapes,
                               is_two_elem_1d(output_size_shape),
                               "output_size must be a 1D input of shape [2]. Got: ",
                                output_size_shape);
```

Reply:
Good catch, `.compatible` was not the correct function to use here. Fixed.